### PR TITLE
[GPU-Plugin] Fix hist and gpu_hist to allow matrices with more than just 2^{32} elements

### PR DESCRIPTION
**Pull request by pseudotensor**
[GPU-Plugin] Fix gpu_hist to allow matrices with more than just 2^{32} elements limited by int (32-bit) and instead use bst_ulong or size_t for element-related quantities.  Also fixed CPU hist algorithm for same reason (uses common code, so had to propagate down through cpu hist algorithm as well). Tested with test_large.py .

### CODE REVIEW

**Comment by pseudotensor** [comment]
@teju85 Can you fix gpu_exact so change int -> size_t or bst_uint -> bst_ulong where appropriate like I did in order to fix the gpu_exact algorithm?  I'm less familiar with it.  Currently gpu_exact won't work for the simple test_large.py python test.

**Comment by pseudotensor** [comment]
@hcho3 Can I go ahead and fix src/common/hist_util.h so that std::vector row_ptr; -> std::vector<size_t> row_ptr; so one can properly handle cases with more than 2 billion matrix elements?  This is a very restrictive limitation currently and numerous data sets have more elements.  Thanks!

File: plugin/updater_gpu/src/device_helpers.cuh [hunk 1]
```diff
@@ -653,11 +653,11 @@ struct BernoulliRng {
 template <typename func_t>
 class LauncherItr {
  public:
-  int idx;
+  size_t idx; // maybe?
```
Comment (pseudotensor) [inline plugin/updater_gpu/src/device_helpers.cuh]: Cleared comment.

File: plugin/updater_gpu/src/device_helpers.cuh [hunk 2]
```diff
@@ -590,7 +590,7 @@ __global__ void launch_n_kernel(int device_idx, size_t begin, size_t end,
 template <int ITEMS_PER_THREAD = 8, int BLOCK_THREADS = 256, typename L>
 inline void launch_n(int device_idx, size_t n, L lambda) {
   safe_cuda(cudaSetDevice(device_idx));
-  const int GRID_SIZE = div_round_up(n, ITEMS_PER_THREAD * BLOCK_THREADS);
+  const int GRID_SIZE = div_round_up(n, ITEMS_PER_THREAD * BLOCK_THREADS); // FIXME: If n is large enough, this limits size can do.  Need to use 2D or 3D grid?
```
Comment (pseudotensor) [inline plugin/updater_gpu/src/device_helpers.cuh]: Ok, I added a note there.  Not a bad limit currently for real data, but can be fixed in future.

**Comment by hcho3** [comment]
@pseudotensor Thanks for the pull request. The use of 32-bit indices was clearly on oversight on our part. Let me review the fast histogram updater for any additional changes necessary.

**Comment by pseudotensor** [comment]
@hcho3 Great, thanks.  Note that there were 2 fundamental issues:
1) the number of elements can be large, so the row_ptr holding where the row is by element count had to be a larger size, size_t or bst_ulong is ok for that, although formally this is not about accessing a pointer.

2) Sometimes the pointer access itself (like an i or local_idx or other things) would index the result from row_ptr or other things whose size could be large, so those quantities had to be upgraded.

3) However, there were only rare cases then access to an array was itself limited.  E.g., while the elements can be large and so the values of row_ptr can be large, the accessor (e.g. i) of row_ptr is not currently large unless we allowed rows themselves to exceed 2-4 billion (which eventually that may be required).  Same for number of features, which is even less likely to exceed 2-4 billion.  I'm sure facebook or google and other such companies have GBM codes that work in that case, maybe lightgbm  by microsoft allows such row counts?  An example where there was a limit is in src/tree/updater_fast_hist.cc when
-      for (unsigned i = gmat.cut->row_ptr[fid]; i < gmat.cut->row_ptr[fid + 1]; ++i) {
-        if (split_pt == gmat.cut->cut[i]) split_cond = static_cast<bst_int>(i);
+      for (size_t i = gmat.cut->row_ptr[fid]; i < gmat.cut->row_ptr[fid + 1]; ++i) {
+        if (split_pt == gmat.cut->cut[i]) split_cond = static_cast<bst_ulong>(i);

Maybe there are other such accessing of arrays that need fixing?

**Comment by khotilov** [comment]
@pseudotensor There's fair amount of code using 32 bit indexing for rows and columns, so the general assumption seemed to be that those would not exceed the limits. The downsides of moving to 64 bit indexing would be, e.g., increased size of sparse data and (it seems like) incompatible data serialization. However, it might be a good idea to parameterize the indexing through a dedicated typedef, so that it could be easily enabled when needed.

**Comment by pseudotensor** [comment]
@khotilov i am only addressing element based accesses to rows, not all row or colums indexing.

**Comment by khotilov** [comment]
@pseudotensor My comment was only WRT you mentioning a possibility to allow #rows or #features to exceed 2-4 billion, not about the substance of this PR (which I agree with).

**Comment by pseudotensor** [comment]
@khotilov  Great, got it.

**Comment by RAMitchell** [comment]
We are not changing any storage types e.g. bst_uint for storing row indices. Memory consumption will be unaffected. Using a typedefs consistently is a good idea.

File: src/common/hist_util.h [hunk 3]
```diff
@@ -60,14 +60,14 @@ struct HistCutUnit {
...
+  HistCutUnit(const bst_float* cut, size_t size)
       : cut(cut), size(size) {}
 };
 
 /*! \brief cut configuration for all the features */
 struct HistCutMatrix {
-  /*! \brief actual unit pointer */
-  std::vector<unsigned> row_ptr;
+  /*! \brief unit pointer to rows by element position */
+  std::vector<size_t> row_ptr;
```
Comment (hcho3) [inline src/common/hist_util.h]: This change is unnecessary -- `HistCutMatrix` is meant to store quantile points for each feature. With a typical configuration of `max_bin=256`, it is unlikely that quantile points would exceed 2^{32} in number.

File: src/common/hist_util.h [hunk 4]
```diff
@@ -60,14 +60,14 @@ struct HistCutUnit {
   // default constructor
   HistCutUnit() {}
   // constructor
-  HistCutUnit(const bst_float* cut, unsigned size)
+  HistCutUnit(const bst_float* cut, size_t size)
```
Comment (hcho3) [inline src/common/hist_util.h]: Same as comment in line 70 -- `unsigned` is okay here

File: src/common/column_matrix.h [hunk 5]
```diff
@@ -86,9 +86,9 @@ class ColumnMatrix {
     type_.resize(nfeature);
     std::fill(feature_counts_.begin(), feature_counts_.end(), 0);
 
-    uint32_t max_val = 0;
+    uint64_t max_val = 0;
```
Comment (hcho3) [inline src/common/column_matrix.h]: This section of code operates on `HistCutMatrix`, which stores the quantiles. So `uint32_t` is fine here.

File: src/common/column_matrix.h [hunk 6]
```diff
@@ -86,9 +86,9 @@ class ColumnMatrix {
...
     XGBOOST_TYPE_SWITCH(this->dtype, {
-      max_val = static_cast<uint32_t>(std::numeric_limits<DType>::max());
+      max_val = static_cast<uint64_t>(std::numeric_limits<size_t>::max());
```
Comment (hcho3) [inline src/common/column_matrix.h]: Same comment as line 89

File: src/common/hist_util.h [hunk 7]
```diff
@@ -120,9 +120,9 @@ struct GHistIndexMatrix {
   inline void GetFeatureCounts(bst_uint* counts) const {
     const unsigned nfeature = cut->row_ptr.size() - 1;
     for (unsigned fid = 0; fid < nfeature; ++fid) {
-      const unsigned ibegin = cut->row_ptr[fid];
-      const unsigned iend = cut->row_ptr[fid + 1];
-      for (unsigned i = ibegin; i < iend; ++i) {
+      const size_t ibegin = cut->row_ptr[fid];
+      const size_t iend = cut->row_ptr[fid + 1];
+      for (size_t i = ibegin; i < iend; ++i) {
```
Comment (hcho3) [inline src/common/hist_util.h]: For lines 123-125, use `unsigned`, as it's related to `HistCutMatrix`

File: src/common/hist_util.h [hunk 8]
```diff
@@ -204,7 +204,7 @@ class HistCollection {
   }
 
   // have we computed a histogram for i-th node?
-  inline bool RowExists(bst_uint nid) const {
+  inline bool RowExists(bst_ulong nid) const {
```
Comment (hcho3) [inline src/common/hist_util.h]: In general, we should stick with `bst_uint` for node id, as the fundamental tree type assumes that we have up to 2^{31} nodes.

File: src/tree/updater_fast_hist.cc [hunk 9]
```diff
@@ -450,11 +450,11 @@ class FastHistMaker: public TreeUpdater {
            choose the column that has a least positive number of discrete bins.
            For dense data (with no missing value),
               the sum of gradient histogram is equal to snode[nid] */
-        const std::vector<unsigned>& row_ptr = gmat.cut->row_ptr;
+        const std::vector<size_t>& row_ptr = gmat.cut->row_ptr;
```
Comment (hcho3) [inline src/tree/updater_fast_hist.cc]: Use `unsigned` here, as it is related to `HistCutMatrix`

File: src/tree/updater_fast_hist.cc [hunk 10]
```diff
@@ -552,8 +552,8 @@ class FastHistMaker: public TreeUpdater {
       bst_int split_cond = -1;
       // convert floating-point split_pt into corresponding bin_id
       // split_cond = -1 indicates that split_pt is less than all known cut points
-      for (unsigned i = gmat.cut->row_ptr[fid]; i < gmat.cut->row_ptr[fid + 1]; ++i) {
-        if (split_pt == gmat.cut->cut[i]) split_cond = static_cast<bst_int>(i);
+      for (size_t i = gmat.cut->row_ptr[fid]; i < gmat.cut->row_ptr[fid + 1]; ++i) {
```
Comment (hcho3) [inline src/tree/updater_fast_hist.cc]: This line and the next operates on quantiles, so let's keep the type as `unsigned`

File: src/tree/updater_fast_hist.cc [hunk 11]
```diff
@@ -786,7 +786,7 @@ class FastHistMaker: public TreeUpdater {
              For dense data (with no missing value),
                 the sum of gradient histogram is equal to snode[nid] */
           GHistRow hist = hist_[nid];
-          const std::vector<unsigned>& row_ptr = gmat.cut->row_ptr;
+          const std::vector<size_t>& row_ptr = gmat.cut->row_ptr;
```
Comment (hcho3) [inline src/tree/updater_fast_hist.cc]: this change not needed

File: src/tree/updater_fast_hist.cc [hunk 12]
```diff
@@ -831,7 +831,7 @@ class FastHistMaker: public TreeUpdater {
       CHECK(d_step == +1 || d_step == -1);
 
       // aliases
-      const std::vector<unsigned>& cut_ptr = gmat.cut->row_ptr;
+      const std::vector<size_t>& cut_ptr = gmat.cut->row_ptr;
```
Comment (hcho3) [inline src/tree/updater_fast_hist.cc]: This change not needed -- related to `HistCutMatrix`

File: src/tree/updater_fast_hist.cc [hunk 13]
```diff
@@ -843,9 +843,9 @@ class FastHistMaker: public TreeUpdater {
       // bin boundaries
       // imin: index (offset) of the minimum value for feature fid
       //       need this for backward enumeration
-      const int imin = cut_ptr[fid];
+      const size_t imin = cut_ptr[fid];
```
Comment (hcho3) [inline src/tree/updater_fast_hist.cc]: This change not needed -- related to `HistCutMatrix`

**Comment by hcho3** [comment]
@pseudotensor I've just gone over the changes. Some comments:
* Overall, I'm happy about the changes
* The `HistCutMatrix` data structure stores quantiles, so `row_ptr` does not need to be `size_t`, despite its name
* On the other hand, `GHistIndexRow` and `GHistIndexMatrix`do store actual matrix elements, so they do need to use `size_t`
* There are some more changes I'd need to make in fast_hist CPU code to accommodate 64-bit indices. Let me upload a diff file for this.

**Comment by pseudotensor** [comment]
@hco3 Great, thanks!  I applied the changes you mentioned.  Feel free to provide a diff on the latest PR and I can apply those too, or you can apply them as a separate PR.  What do you think is best?

**Comment by hcho3** [comment]
@pseudotensor I've made some changes on the CPU algorithm. For the sake of consistency, I use `size_t` for row indices and `uint32_t` for quantile indices.

[patch.txt](https://github.com/dmlc/xgboost/files/1151448/patch.txt)

**Comment by pseudotensor** [comment]
Hi @hcho3  The tests segfault when reaching hist when I applied the changes in your comments or with also the patch.txt.  https://travis-ci.org/dmlc/xgboost/jobs/254273688

**Comment by pseudotensor** [comment]
Hi @hcho3, With appveryor, I got 5 errors for windows related to OpenMP requiring loop variable to be signed.  Apart from that, I just applied your changes.

**Comment by hcho3** [comment]
@pseudotensor Thanks! As for OpenMP, it appears to be a convention throughout xgboost to use the typedef `bst_omp_uint`, which is actually signed on Windows platform. Can you apply [this patch](https://github.com/dmlc/xgboost/files/1152037/patch2.txt) to be consistent with the convention?

**Comment by pseudotensor** [comment]
@hcho3 Great, I applied that patch2.txt

**Pull request merged**